File: tests/test_state_machine.py
"""Test app/state_machine.py — referensi murni logika transisi misi.

Memvalidasi keputusan transisi yang diekstrak dari run() di navigator.py.
Skenario meniru perilaku kapal nyata (nilai threshold datang dari
telemetri/vision/timer — bukan konstanta di sini).

Jalankan:  python3 -m unittest discover -s tests -v
"""

import unittest

from app.state_machine import (
    APPROACH_BLUE_BOX_ALIGN,
    APPROACH_BLUE_BOX_SEARCH,
    APPROACH_BOX_ALIGN,
    APPROACH_BOX_SEARCH,
    APPROACH_RED_BOX_ALIGN,
    APPROACH_RED_BOX_SEARCH,
    BLUE_BOX_RETREAT,
    MISSION_COMPLETE,
    NO_TELEM,
    NO_WAYPOINTS,
    RED_BOX_DOCKED,
    RETREAT,
    SR_GOTO_NEUTRAL,
    SR_IDLE,
    SR_START_BRAKE,
    SR_START_REVERSE,
    TAKE_BLUE_BOX_PHOTO,
    TAKE_PHOTO,
    TAKE_WAYPOINT_PHOTO,
    WAYPOINT_NAV,
    WAYPOINT_TRANSITION,
    SmInput,
    sm_guard,
    sm_transition,
)


def _nav(wp=0, n=5, **kw):
    return SmInput(wp_idx=wp, n_wp=n, **kw)


class TestGuard(unittest.TestCase):
    def test_tanpa_telemetri(self):
        self.assertEqual(sm_guard(0, 1, 1, 0, 5), NO_TELEM)
        self.assertEqual(sm_guard(1, 0, 1, 0, 5), NO_TELEM)

    def test_tanpa_waypoint(self):
        self.assertEqual(sm_guard(1, 1, 0, 0, 5), NO_WAYPOINTS)

    def test_misi_habis(self):
        self.assertEqual(sm_guard(1, 1, 1, 5, 5), MISSION_COMPLETE)

    def test_boleh_lanjut(self):
        self.assertIsNone(sm_guard(1, 1, 1, 4, 5))


class TestNavGate(unittest.TestCase):
    def test_nav_biasa_tetap(self):
        r = sm_transition(WAYPOINT_NAV, SR_IDLE, _nav(at_waypoint=0))
        self.assertEqual(r.next_state, WAYPOINT_NAV)
        self.assertEqual(r.wp_inc, 0)

    def test_sampai_wp_lanjut_transisi(self):
        r = sm_transition(WAYPOINT_NAV, SR_IDLE, _nav(at_waypoint=1))
        self.assertEqual(r.next_state, WAYPOINT_TRANSITION)
        self.assertEqual(r.wp_inc, 1)
        self.assertEqual(r.reset_task_timer, 1)
        self.assertEqual(r.reset_vision, 1)

    def test_wp_terakhir_selesai(self):
        r = sm_transition(WAYPOINT_NAV, SR_IDLE,
                          _nav(wp=4, at_waypoint=1))
        self.assertEqual(r.next_state, MISSION_COMPLETE)
        self.assertEqual(r.wp_inc, 1)

    def test_transisi_menunggu_waktu(self):
        r = sm_transition(WAYPOINT_TRANSITION, SR_IDLE,
                          SmInput(transition_over=0))
        self.assertEqual(r.next_state, WAYPOINT_TRANSITION)
        r = sm_transition(WAYPOINT_TRANSITION, SR_IDLE,
                          SmInput(transition_over=1))
        self.assertEqual(r.next_state, WAYPOINT_NAV)

    def test_foto_wp_stop(self):
        r = sm_transition(WAYPOINT_NAV, SR_IDLE,
                          _nav(wp=2, at_waypoint=1, wp_in_stop_and_photo=1))
        self.assertEqual(r.next_state, TAKE_WAYPOINT_PHOTO)
        self.assertEqual(r.reset_task_timer, 1)
        self.assertEqual(r.wp_inc, 0)

    def test_foto_wp_lanjut_atau_docking(self):
        r = sm_transition(TAKE_WAYPOINT_PHOTO, SR_IDLE,
                          _nav(wp=3, wp_photo_over=1,
                               wp_is_red_after=1, red_model=1))
        self.assertEqual(r.next_state, APPROACH_RED_BOX_SEARCH)
        self.assertNotEqual(r.next_state, WAYPOINT_TRANSITION)  # docking dulu
        r = sm_transition(TAKE_WAYPOINT_PHOTO, SR_IDLE,
                          _nav(wp=1, wp_photo_over=1))
        self.assertEqual(r.next_state, WAYPOINT_TRANSITION)
        self.assertEqual(r.wp_inc, 1)


class TestMisiFotoBox(unittest.TestCase):
    def test_urutan_full(self):
        steps = [
            _nav(wp=0, wp_in_photo_box_legs=1, green_model=1),
            SmInput(box_found=1, box_confirmed=0),
            SmInput(box_found=1, box_confirmed=1),
            SmInput(box_found=1, box_aligned=0),
            SmInput(box_found=1, box_aligned=1),
            SmInput(),
            SmInput(brake_over=1),
            SmInput(neutral_over=1),
            SmInput(retreat_over=0),
            SmInput(retreat_over=1),
        ]
        # states[i] = state BARU sesudah langkah ke-i (state awal WAYPOINT_NAV
        # tidak ikut dicek — dicek lewat hasil transisi).
        states = [APPROACH_BOX_SEARCH, APPROACH_BOX_SEARCH,
                  APPROACH_BOX_ALIGN, APPROACH_BOX_ALIGN, TAKE_PHOTO,
                  RETREAT, RETREAT, RETREAT, RETREAT, WAYPOINT_NAV]
        retreat = [SR_IDLE, SR_IDLE, SR_IDLE, SR_IDLE, SR_IDLE,
                   SR_START_BRAKE, SR_GOTO_NEUTRAL, SR_START_REVERSE,
                   SR_START_REVERSE, SR_IDLE]
        state = WAYPOINT_NAV
        step = SR_IDLE
        for i, inp in enumerate(steps):
            r = sm_transition(state, step, inp)
            state = r.next_state
            step = r.retreat_step
            self.assertEqual(r.next_state, states[i], f"langkah {i}")
            self.assertEqual(r.retreat_step, retreat[i], f"step {i}")

    def test_box_hilang_2s_ambil_foto(self):
        r = sm_transition(APPROACH_BOX_ALIGN, SR_IDLE,
                          SmInput(box_found=0, box_lost_over_2s=1))
        self.assertEqual(r.next_state, TAKE_PHOTO)


class TestMisiBirudanDocking(unittest.TestCase):
    def test_urutan_full(self):
        steps = [
            _nav(wp=0, n=10, wp_in_blue_box_legs=1, blue_model=1),
            SmInput(blue_found=1, blue_confirmed=1),
            SmInput(blue_found=1, blue_aligned=1),
            SmInput(blue_photo_over=0),
            SmInput(blue_photo_over=1),
            SmInput(brake_over=1),
            SmInput(neutral_over=1),
            SmInput(retreat_over=1, red_model=1),
            SmInput(red_found=0),
            SmInput(red_found=1),
            SmInput(red_found=1, red_aligned=0),
            SmInput(red_found=1, red_aligned=1),
            SmInput(dock_hold_over=0),
            SmInput(dock_hold_over=1),
        ]
        states = [APPROACH_BLUE_BOX_SEARCH, APPROACH_BLUE_BOX_ALIGN,
                  TAKE_BLUE_BOX_PHOTO, TAKE_BLUE_BOX_PHOTO,
                  BLUE_BOX_RETREAT, BLUE_BOX_RETREAT, BLUE_BOX_RETREAT,
                  APPROACH_RED_BOX_SEARCH, APPROACH_RED_BOX_SEARCH,
                  APPROACH_RED_BOX_ALIGN, APPROACH_RED_BOX_ALIGN,
                  RED_BOX_DOCKED, RED_BOX_DOCKED, WAYPOINT_TRANSITION]
        state = WAYPOINT_NAV
        step = SR_IDLE
        wp_idx = 0
        for i, inp in enumerate(steps):
            inp.wp_idx = wp_idx  # snapshot wp selalu mengikuti kondisi saat ini
            if inp.n_wp == 0:
                inp.n_wp = 10
            r = sm_transition(state, step, inp)
            state = r.next_state
            step = r.retreat_step
            wp_idx += r.wp_inc
            self.assertEqual(r.next_state, states[i], f"langkah {i}")

    def test_blue_retreat_tanpa_model_merah(self):
        r = sm_transition(BLUE_BOX_RETREAT, SR_START_REVERSE,
                          SmInput(retreat_over=1, red_model=0))
        self.assertEqual(r.next_state, WAYPOINT_NAV)
        self.assertEqual(r.wp_inc, 1)


class TestRedDock(unittest.TestCase):
    def test_red_lost_kembali_search(self):
        r = sm_transition(APPROACH_RED_BOX_ALIGN, SR_IDLE,
                          SmInput(red_found=0))
        self.assertEqual(r.next_state, APPROACH_RED_BOX_SEARCH)

    def test_red_align_docked(self):
        r = sm_transition(APPROACH_RED_BOX_ALIGN, SR_IDLE,
                          SmInput(red_found=1, red_aligned=1))
        self.assertEqual(r.next_state, RED_BOX_DOCKED)
        self.assertEqual(r.reset_task_timer, 1)

    def test_dock_selesai_lanjut(self):
        r = sm_transition(RED_BOX_DOCKED, SR_IDLE,
                          SmInput(wp_idx=0, n_wp=10, dock_hold_over=1))
        self.assertEqual(r.next_state, WAYPOINT_TRANSITION)
        self.assertEqual(r.wp_inc, 1)


class TestStateTanpaTransisi(unittest.TestCase):
    def test_guard_states_tetap(self):
        for state in (MISSION_COMPLETE, "WAITING_GPS", "NO_TELEM",
                      "NO_WAYPOINTS"):
            r = sm_transition(state, SR_IDLE, SmInput())
            self.assertEqual(r.next_state, state)
            self.assertEqual(r.wp_inc, 0)


if __name__ == "__main__":
    unittest.main()

File: core/tests/test_state_machine.c
/*
 * core/tests/test_state_machine.c — Unit test mesin state misi (C)
 *
 * Skenario nilai emas = trace urutan state yang DIHARAPKAN dari perilaku
 * run() di navigator.py (diekstrak ke app/state_machine.py, C wajib sama).
 * Skenario:
 *   A. Gate nav murni: WP_NAV -> TRANSITION -> WP_NAV ... -> MISSION_COMPLETE
 *   B. Foto box hijau penuh + sub-machine mundur (brake->netral->mundur)
 *   C. Foto box biru + docking red box + hold + lanjut misi
 *   D. Jalan buntu: target hilang saat ALIGN, RED lost, retreat error
 *   E. Guard loop utama (NO_TELEM / NO_WAYPOINTS / MISSION_COMPLETE)
 *   F. Stop-and-photo waypoint + red-dock-setelah-WP
 *   G. Konversi nama <-> enum
 *
 * Jalankan:
 *   gcc -I core/include core/tests/test_state_machine.c core/src/state_machine.c \
 *       -o /tmp/opencode/test_sm && /tmp/opencode/test_sm
 */
#include "state_machine.h"

#include <stdio.h>
#include <string.h>

static int failures = 0;

/* Cek satu transisi; cetak nama state biar mudah dibaca. */
static void expect(const char *tag, sm_state_t state, sm_retreat_step_t step,
                   const sm_input_t *in,
                   sm_state_t want_next, int want_wp_inc,
                   int want_reset_tt, int want_reset_vis,
                   sm_retreat_step_t want_step) {
    const sm_result_t r = sm_transition(state, step, in);
    int bad = 0;
    if (r.next != want_next) {
        printf("FAIL %s: next=%s harap=%s\n", tag, sm_state_name(r.next),
               sm_state_name(want_next));
        bad = 1;
    }
    if (r.wp_inc != want_wp_inc) {
        printf("FAIL %s: wp_inc=%d harap=%d\n", tag, r.wp_inc, want_wp_inc);
        bad = 1;
    }
    if (r.reset_task_timer != want_reset_tt) {
        printf("FAIL %s: reset_task_timer=%d harap=%d\n", tag,
               r.reset_task_timer, want_reset_tt);
        bad = 1;
    }
    if (r.reset_vision != want_reset_vis) {
        printf("FAIL %s: reset_vision=%d harap=%d\n", tag,
               r.reset_vision, want_reset_vis);
        bad = 1;
    }
    if (r.retreat_step != want_step) {
        printf("FAIL %s: retreat_step=%s harap=%s\n", tag,
               sm_retreat_step_name(r.retreat_step),
               sm_retreat_step_name(want_step));
        bad = 1;
    }
    if (!bad) {
        printf("ok    %-38s %s -> %s\n", tag, sm_state_name(state),
               sm_state_name(r.next));
    } else {
        failures++;
    }
}

static void expect_stay(const char *tag, sm_state_t state,
                        sm_retreat_step_t step, const sm_input_t *in,
                        sm_retreat_step_t want_step) {
    expect(tag, state, step, in, state, 0, 0, 0, want_step);
}

int main(void) {
    /* ---------------------- A. Gate nav murni ---------------------- */
    expect("A1 nav->transition", SM_WAYPOINT_NAV, SM_STEP_IDLE,
           &(sm_input_t){ .wp_idx = 0, .n_wp = 5, .at_waypoint = 1 },
           SM_WAYPOINT_TRANSITION, 1, 1, 1, SM_STEP_IDLE);
    expect_stay("A2 transition menunggu", SM_WAYPOINT_TRANSITION, SM_STEP_IDLE,
                &(sm_input_t){ .transition_over = 0 }, SM_STEP_IDLE);
    expect("A3 transition selesai", SM_WAYPOINT_TRANSITION, SM_STEP_IDLE,
           &(sm_input_t){ .transition_over = 1 },
           SM_WAYPOINT_NAV, 0, 0, 0, SM_STEP_IDLE);
    expect("A4 wp terakhir -> selesai", SM_WAYPOINT_NAV, SM_STEP_IDLE,
           &(sm_input_t){ .wp_idx = 4, .n_wp = 5, .at_waypoint = 1 },
           SM_MISSION_COMPLETE, 1, 0, 1, SM_STEP_IDLE);
    expect_stay("A5 nav biasa", SM_WAYPOINT_NAV, SM_STEP_IDLE,
                &(sm_input_t){ .wp_idx = 2, .n_wp = 5, .at_waypoint = 0 },
                SM_STEP_IDLE);

    /* ----------------- B. Foto box hijau penuh ----------------- */
    expect("B1 leg foto-box", SM_WAYPOINT_NAV, SM_STEP_IDLE,
           &(sm_input_t){ .wp_idx = 0, .n_wp = 5,
                          .wp_in_photo_box_legs = 1, .green_model = 1 },
           SM_APPROACH_BOX_SEARCH, 0, 0, 1, SM_STEP_IDLE);
    expect_stay("B2 search verifikasi", SM_APPROACH_BOX_SEARCH, SM_STEP_IDLE,
                &(sm_input_t){ .box_found = 1, .box_confirmed = 0 },
                SM_STEP_IDLE);
    expect("B3 search terkunci", SM_APPROACH_BOX_SEARCH, SM_STEP_IDLE,
           &(sm_input_t){ .box_found = 1, .box_confirmed = 1 },
           SM_APPROACH_BOX_ALIGN, 0, 0, 1, SM_STEP_IDLE);
    expect_stay("B4 align maju", SM_APPROACH_BOX_ALIGN, SM_STEP_IDLE,
                &(sm_input_t){ .box_found = 1, .box_aligned = 0 },
                SM_STEP_IDLE);
    expect("B5 align sampai", SM_APPROACH_BOX_ALIGN, SM_STEP_IDLE,
           &(sm_input_t){ .box_found = 1, .box_aligned = 1 },
           SM_TAKE_PHOTO, 0, 0, 1, SM_STEP_IDLE);
    expect("B6 take photo -> retreat", SM_TAKE_PHOTO, SM_STEP_IDLE,
           &(sm_input_t){},
           SM_RETREAT, 0, 1, 0, SM_STEP_START_BRAKE);
    expect_stay("B7 brake", SM_RETREAT, SM_STEP_START_BRAKE,
                &(sm_input_t){ .brake_over = 0 }, SM_STEP_START_BRAKE);
    expect("B8 brake selesai", SM_RETREAT, SM_STEP_START_BRAKE,
           &(sm_input_t){ .brake_over = 1 },
           SM_RETREAT, 0, 0, 0, SM_STEP_GOTO_NEUTRAL);
    expect_stay("B9 netral", SM_RETREAT, SM_STEP_GOTO_NEUTRAL,
                &(sm_input_t){ .neutral_over = 0 }, SM_STEP_GOTO_NEUTRAL);
    expect("B10 netral selesai", SM_RETREAT, SM_STEP_GOTO_NEUTRAL,
           &(sm_input_t){ .neutral_over = 1 },
           SM_RETREAT, 0, 1, 0, SM_STEP_START_REVERSE);
    expect_stay("B11 mundur", SM_RETREAT, SM_STEP_START_REVERSE,
                &(sm_input_t){ .retreat_over = 0 }, SM_STEP_START_REVERSE);
    expect("B12 mundur selesai", SM_RETREAT, SM_STEP_START_REVERSE,
           &(sm_input_t){ .retreat_over = 1 },
           SM_WAYPOINT_NAV, 1, 0, 1, SM_STEP_IDLE);

    /* -------------- C. Box biru + docking merah ------------- */
    expect("C1 leg foto-biru", SM_WAYPOINT_NAV, SM_STEP_IDLE,
           &(sm_input_t){ .wp_idx = 0, .n_wp = 10,
                          .wp_in_blue_box_legs = 1, .blue_model = 1 },
           SM_APPROACH_BLUE_BOX_SEARCH, 0, 0, 1, SM_STEP_IDLE);
    expect("C2 biru search terkunci", SM_APPROACH_BLUE_BOX_SEARCH, SM_STEP_IDLE,
           &(sm_input_t){ .blue_found = 1, .blue_confirmed = 1 },
           SM_APPROACH_BLUE_BOX_ALIGN, 0, 0, 1, SM_STEP_IDLE);
    expect("C3 biru align sampai", SM_APPROACH_BLUE_BOX_ALIGN, SM_STEP_IDLE,
           &(sm_input_t){ .blue_found = 1, .blue_aligned = 1 },
           SM_TAKE_BLUE_BOX_PHOTO, 0, 1, 1, SM_STEP_IDLE);
    expect_stay("C4 biru fotokan", SM_TAKE_BLUE_BOX_PHOTO, SM_STEP_IDLE,
                &(sm_input_t){ .blue_photo_over = 0 }, SM_STEP_IDLE);
    expect("C5 biru foto selesai", SM_TAKE_BLUE_BOX_PHOTO, SM_STEP_IDLE,
           &(sm_input_t){ .blue_photo_over = 1 },
           SM_BLUE_BOX_RETREAT, 0, 1, 0, SM_STEP_START_BRAKE);
    expect("C6 biru brake selesai", SM_BLUE_BOX_RETREAT, SM_STEP_START_BRAKE,
           &(sm_input_t){ .brake_over = 1 },
           SM_BLUE_BOX_RETREAT, 0, 0, 0, SM_STEP_GOTO_NEUTRAL);
    expect("C7 biru netral selesai", SM_BLUE_BOX_RETREAT, SM_STEP_GOTO_NEUTRAL,
           &(sm_input_t){ .neutral_over = 1 },
           SM_BLUE_BOX_RETREAT, 0, 1, 0, SM_STEP_START_REVERSE);
    expect("C8 biru mundur -> docking", SM_BLUE_BOX_RETREAT,
           SM_STEP_START_REVERSE,
           &(sm_input_t){ .retreat_over = 1, .red_model = 1 },
           SM_APPROACH_RED_BOX_SEARCH, 0, 0, 1, SM_STEP_IDLE);
    expect_stay("C9 red search rotasi", SM_APPROACH_RED_BOX_SEARCH,
                SM_STEP_IDLE, &(sm_input_t){ .red_found = 0 }, SM_STEP_IDLE);
    expect("C10 red search ketemu", SM_APPROACH_RED_BOX_SEARCH, SM_STEP_IDLE,
           &(sm_input_t){ .red_found = 1 },
           SM_APPROACH_RED_BOX_ALIGN, 0, 0, 1, SM_STEP_IDLE);
    expect_stay("C11 red align maju", SM_APPROACH_RED_BOX_ALIGN, SM_STEP_IDLE,
                &(sm_input_t){ .red_found = 1, .red_aligned = 0 },
                SM_STEP_IDLE);
    expect("C12 red dock", SM_APPROACH_RED_BOX_ALIGN, SM_STEP_IDLE,
           &(sm_input_t){ .red_found = 1, .red_aligned = 1 },
           SM_RED_BOX_DOCKED, 0, 1, 0, SM_STEP_IDLE);
    expect_stay("C13 dock hold", SM_RED_BOX_DOCKED, SM_STEP_IDLE,
                &(sm_input_t){ .dock_hold_over = 0 }, SM_STEP_IDLE);
    expect("C14 dock selesai -> lanjut", SM_RED_BOX_DOCKED, SM_STEP_IDLE,
           &(sm_input_t){ .wp_idx = 0, .n_wp = 10, .dock_hold_over = 1 },
           SM_WAYPOINT_TRANSITION, 1, 1, 1, SM_STEP_IDLE);
    expect("C15 dock selesai -> complete", SM_RED_BOX_DOCKED, SM_STEP_IDLE,
           &(sm_input_t){ .wp_idx = 9, .n_wp = 10, .dock_hold_over = 1 },
           SM_MISSION_COMPLETE, 1, 0, 1, SM_STEP_IDLE);

    /* ----------------- D. Jalan buntu / hilang ----------------- */
    expect_stay("D1 box hilang < 2s", SM_APPROACH_BOX_ALIGN, SM_STEP_IDLE,
                &(sm_input_t){ .box_found = 0, .box_lost_over_2s = 0 },
                SM_STEP_IDLE);
    expect("D2 box hilang > 2s", SM_APPROACH_BOX_ALIGN, SM_STEP_IDLE,
           &(sm_input_t){ .box_found = 0, .box_lost_over_2s = 1 },
           SM_TAKE_PHOTO, 0, 0, 1, SM_STEP_IDLE);
    expect("D3 biru hilang > 2s", SM_APPROACH_BLUE_BOX_ALIGN, SM_STEP_IDLE,
           &(sm_input_t){ .blue_found = 0, .blue_lost_over_2s = 1 },
           SM_TAKE_BLUE_BOX_PHOTO, 0, 1, 1, SM_STEP_IDLE);
    expect("D4 red hilang -> search", SM_APPROACH_RED_BOX_ALIGN, SM_STEP_IDLE,
           &(sm_input_t){ .red_found = 0 },
           SM_APPROACH_RED_BOX_SEARCH, 0, 0, 0, SM_STEP_IDLE);
    expect("D5 biru mundur tanpa model merah", SM_BLUE_BOX_RETREAT,
           SM_STEP_START_REVERSE,
           &(sm_input_t){ .wp_idx = 3, .n_wp = 10,
                          .retreat_over = 1, .red_model = 0 },
           SM_WAYPOINT_NAV, 1, 0, 1, SM_STEP_IDLE);
    expect("D6 retreat error step", SM_RETREAT, SM_STEP_IDLE,
           &(sm_input_t){},
           SM_WAYPOINT_NAV, 0, 0, 0, SM_STEP_IDLE);

    /* ------------------ E. Guard loop utama ------------------ */
    if (sm_guard(0, 1, 1, 0, 5) != SM_NO_TELEM) {
        printf("FAIL E1 guard GPS off\n"); failures++;
    } else { printf("ok    E1 guard no-GPS           = NO_TELEM\n"); }
    if (sm_guard(1, 0, 1, 0, 5) != SM_NO_TELEM) {
        printf("FAIL E2 guard yaw off\n"); failures++;
    } else { printf("ok    E2 guard no-yaw           = NO_TELEM\n"); }
    if (sm_guard(1, 1, 0, 0, 5) != SM_NO_WAYPOINTS) {
        printf("FAIL E3 guard no-waypoints\n"); failures++;
    } else { printf("ok    E3 guard no-waypoints     = NO_WAYPOINTS\n"); }
    if (sm_guard(1, 1, 1, 5, 5) != SM_MISSION_COMPLETE) {
        printf("FAIL E4 guard done\n"); failures++;
    } else { printf("ok    E4 guard all-done         = MISSION_COMPLETE\n"); }
    if (sm_guard(1, 1, 1, 4, 5) != SM_NONE) {
        printf("FAIL E5 guard continue\n"); failures++;
    } else { printf("ok    E5 guard boleh-lanjut      = SM_NONE\n"); }

    /* ------------- F. Stop-and-photo + red setelah WP ------------- */
    expect("F1 stop-photo", SM_WAYPOINT_NAV, SM_STEP_IDLE,
           &(sm_input_t){ .wp_idx = 2, .n_wp = 5, .at_waypoint = 1,
                          .wp_in_stop_and_photo = 1 },
           SM_TAKE_WAYPOINT_PHOTO, 0, 1, 0, SM_STEP_IDLE);
    expect_stay("F2 foto waypoint", SM_TAKE_WAYPOINT_PHOTO, SM_STEP_IDLE,
                &(sm_input_t){ .wp_photo_over = 0 }, SM_STEP_IDLE);
    expect("F3 foto wp -> docking", SM_TAKE_WAYPOINT_PHOTO, SM_STEP_IDLE,
           &(sm_input_t){ .wp_idx = 3, .n_wp = 5, .wp_photo_over = 1,
                          .wp_is_red_after = 1, .red_model = 1 },
           SM_APPROACH_RED_BOX_SEARCH, 0, 0, 1, SM_STEP_IDLE);
    expect("F4 foto wp -> lanjut", SM_TAKE_WAYPOINT_PHOTO, SM_STEP_IDLE,
           &(sm_input_t){ .wp_idx = 1, .n_wp = 5, .wp_photo_over = 1 },
           SM_WAYPOINT_TRANSITION, 1, 1, 1, SM_STEP_IDLE);
    expect("F5 foto wp terakhir", SM_TAKE_WAYPOINT_PHOTO, SM_STEP_IDLE,
           &(sm_input_t){ .wp_idx = 4, .n_wp = 5, .wp_photo_over = 1 },
           SM_MISSION_COMPLETE, 1, 0, 1, SM_STEP_IDLE);

    /* ------------------ G. Konversi nama <-> enum ------------------ */
    if (strcmp(sm_state_name(SM_RED_BOX_DOCKED), "RED_BOX_DOCKED") != 0) {
        printf("FAIL G1 nama state\n"); failures++;
    } else { printf("ok    G1 sm_state_name          = RED_BOX_DOCKED\n"); }
    if (sm_state_from_name("WAYPOINT_NAV") != SM_WAYPOINT_NAV) {
        printf("FAIL G2 from-name\n"); failures++;
    } else { printf("ok    G2 sm_state_from_name      = WAYPOINT_NAV\n"); }
    if (sm_state_from_name("TIDAK_ADA") != SM_NONE) {
        printf("FAIL G3 from-name tak dikenal\n"); failures++;
    } else { printf("ok    G3 nama tak dikenal        = SM_NONE\n"); }
    if (strcmp(sm_retreat_step_name(SM_STEP_GOTO_NEUTRAL),
               "GOTO_NEUTRAL") != 0) {
        printf("FAIL G4 nama step\n"); failures++;
    } else { printf("ok    G4 sm_retreat_step_name     = GOTO_NEUTRAL\n"); }

    if (failures == 0) {
        printf("\nSEMUA skenario state_machine LULUS (C == app/state_machine.py)\n");
        return 0;
    }
    printf("\n%d kasus GAGAL\n", failures);
    return 1;
}